### PR TITLE
WPF support

### PR DESCRIPTION
**Pull request by wjk**
This PR adds support for WPF-based projects to the new, SDK-based project system, including full designer support. It should support C#, F#, and VB.NET projects (although I have only tested it with C#).

Note that for each project that wants to include WPF code, the following code must be placed into its project file:

```xml
<PropertyGroup>
  <EnableWpfTargets>true</EnableWpfTargets>
</PropertyGroup>
```

This is because I didn't want the WPF targets being included where they are not needed, because they are fairly invasive. If this code is not added, the WPF targets will not be included (and weird build breakage will likely result).

This PR supersedes #20, which should be closed once this is merged. Thanks!

### CODE REVIEW

**Comment by wjk** [comment]
Note that you’ll have to add `**/*.xml` to [the `<None />` line in the csproj](https://github.com/onovotny/MSBuildSdkExtras/blob/master/MSBuild.Sdk.Extras/MSBuild.Sdk.Extras.csproj#L27), or else the MSBuild rules files won‘t get included in the NuGet package, and weird failures will result. I’d do this myself, but it’s getting late. 😄 I can either add that code tomorrow, or you can do it yourself once this is merged. Thanks!

**Comment by davkean** [comment]
Can we contribute the built-in item rules to http://github.com/dotnet/project-system instead of adding them to this project? Rules should only be added for items that you own yourself. I will happily take the contribution.

I would definitely recommend you run this targets file under the new evaluation profiler: https://github.com/Microsoft/msbuild/pull/2765. Globs and updates are not free - and this adds a bunch. Slow evaluation leads to increase project load item, larger design-time builds for those projects that reference you (including .NET Framework based on the legacy project system), which can lead to UI delays and slow project manipulation.

File: MSBuild.Sdk.Extras/build/platforms/languageTargets/Wpf.targets [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    <LanguageTargets>$(MSBuildExtensionsPath)\$(VisualStudioVersion)\Bin\Microsoft.$(_SdkLanguageName).targets</LanguageTargets>
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <EnableDefaultWpfItems Condition="'$(EnableDefaultWpfItems)' == ''">True</EnableDefaultWpfItems>
+    <DefaultWpfPageExcludes>$(DefaultWpfPageExcludes);$(MSBuildProjectDirectory)/App.xaml</DefaultWpfPageExcludes>
+  </PropertyGroup>
+
+  <ItemGroup Condition="'$(EnableDefaultItems)' == 'True'">
+    <None Remove="**/*.xaml" Condition="'$(EnableDefaultWpfItems)' == 'True'" />
```
Comment (onovotny) [inline MSBuild.Sdk.Extras/build/platforms/languageTargets/Wpf.targets]: The removes can have an extra condition:
https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.props#L29
` And '$(EnableDefaultNoneItems)' == 'true`

File: MSBuild.Sdk.Extras/build/platforms/languageTargets/Wpf.targets [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    <Page Include="**/*.xaml" Exclude="$(DefaultItemExcludes);$(DefaultExcludesInProjectFolder);$(DefaultWpfPageExcludes)" Condition="'$(EnableDefaultWpfItems)' == 'True'" />
+    <None Remove="$(MSBuildProjectDirectory)/App.xaml" Condition="'$(EnableDefaultWpfItems)' == 'True' and Exists('$(MSBuildProjectDirectory)/App.xaml')" />
+    <ApplicationDefinition Include="$(MSBuildProjectDirectory)/App.xaml" Condition="'$(EnableDefaultWpfItems)' == 'True' and Exists('$(MSBuildProjectDirectory)/App.xaml')" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <Page Update="@(Page)" SubType="Designer" Generator="MSBuild:Compile" />
```
Comment (onovotny) [inline MSBuild.Sdk.Extras/build/platforms/languageTargets/Wpf.targets]: Can we do this without an update? Just put this on the Page include above?

File: MSBuild.Sdk.Extras/build/platforms/languageTargets/Wpf.targets [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    <Compile Update="**\*.xaml.cs" SubType="Designer" DependentUpon="%(Filename)" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <PropertyPageSchema Include="$(MSBuildThisFileDirectory)Rules/Wpf.ProjectItemsSchema.xml">
```
Comment (onovotny) [inline MSBuild.Sdk.Extras/build/platforms/languageTargets/Wpf.targets]: @davkean I'm going to accept this here for now. I really wish there was a VS Version property passed into the design time build that had the full version info. Closest I see now is 15.0, which isn't good enough.

I want to get these in ASAP but would prefer to condition them on specific builds of VS so we could use the in-box ones.

**Comment by wjk** [comment]
@davkean How should I run the MSBuild profiler, and how should I interpret the results? This is my first time using this feature. Thanks!

**Comment by onovotny** [comment]
@wjk @sharwell I'm seeing errors while trying to use this as a package.

https://github.com/onovotny/MSBuildSdkExtras/tree/wpf-sample

It seems to work in the test projects when they're directly included. Any ideas?

**Comment by onovotny** [comment]
There's a start of workaround that I added to the branch, but I still get this:

```
1>Target GenerateTemporaryTargetAssembly:
1>  Target CoreCompile:
1>    Using shared compilation with compiler from directory: C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\Roslyn
1>    MainWindow.xaml.cs(25,13,25,32): error CS0103: The name 'InitializeComponent' does not exist in the current context
1>  Done building target "CoreCompile" in project "jyurx4g3.tmp_proj" -- FAILED.
```
From what I can tell looking at the binlog, it's not generating the file for the XAML files.

**Comment by wjk** [comment]
@onovotny Yeah, that's what I'm seeing as well. I'm not seeing the `MainWindow.g.cs` file anywhere. It _is_ generating the `App.g.cs` file, though. How weird.

**Comment by onovotny** [comment]
Wait, nm, I had an error...was missing the `x:Class`.

**Comment by onovotny** [comment]
Still have a nasty work around for https://github.com/NuGet/Home/issues/5894 :(

**Comment by wjk** [comment]
Yep, that was it re the `MainWindow.g.cs` issue. Will submit a second PR re the NuGet bug using material from Microsoft/msbuild#2701. Thanks!

**Comment by wjk** [comment]
Actually, as it turns out, the linked MSBuild issue depends on another PR (NuGet/NuGet.Client#1786), which may take a while to merge because of licensing issues (they say it contains some code that hasn't gotten the official OK from MS Legal). Nevermind my previous comment, as it wouldn't have worked.

**Comment by sharwell** [comment]
> Still have a nasty work around for NuGet/Home#5894 :(

Can you point out the hack?

**Comment by onovotny** [comment]
@sharwell The fact the csproj has to have this at the top: https://github.com/NuGetPackageExplorer/NuGetPackageExplorer/blob/master/PackageExplorer/NuGetPackageExplorer.csproj#L3-L7

and this at the bottom: https://github.com/NuGetPackageExplorer/NuGetPackageExplorer/blob/master/PackageExplorer/NuGetPackageExplorer.csproj#L50-L51

**Comment by onovotny** [comment]
@sharwell It's because `PackageReference` for WPF projects is currently broken in general due to the two-pass compile and the `tmp_proj` file extension it uses.

**Comment by mungojam** [comment]
I tried using the extras package plus the workaround yesterday and got 2 errors:

1. It couldn't find a main method in the WPF project 
2. It couldn't resolve the InitializeComponent method. That was after using the https://github.com/hvanbakel/CsprojToVs2017 project converter

For now, I've reverted to using the old style project. Let me know if you think it should work and I can try again and provide more details.

**Pull request merged**